### PR TITLE
WAR load support

### PR DESCRIPTION
**Pull request by mgudemann**
This adds support to load class files from web application archive (WAR) files.
requires #604

### CODE REVIEW

File: regression/cbmc-java/war-file1/url.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+https://tomcat.apache.org/tomcat-6.0-doc/appdev/sample/
```
Comment (tautschnig) [inline regression/cbmc-java/war-file1/url.txt]: I very much appreciate that comment - but you might even place it in the test.desc file (so as not to create a non-standard file): you can place comments after another `--`, see `test.pl --help`

File: src/java_bytecode/jar_file.cpp [hunk 2]
```diff
@@ -77,6 +78,11 @@ void jar_filet::open(
       assert(filename_length==filename_len);
       std::string file_name(filename_char);
 
+      // remove WAR class prefix
+      const std::string war_class_prefix("WEB-INF/classes/");
```
Comment (tautschnig) [inline src/java_bytecode/jar_file.cpp]: Would that also be the prefix on Windows?

File: src/java_bytecode/jar_file.cpp [hunk 3]
```diff
@@ -48,8 +77,27 @@ void jar_filet::open(const std::string &filename)
         mz_zip_reader_get_filename(&zip, i, filename_char, filename_length);
       assert(filename_length==filename_len);
       std::string file_name(filename_char);
+
+      // remove WAR class prefix
+      const std::string war_class_prefix("WEB-INF/classes/");
```
Comment (tautschnig) [inline src/java_bytecode/jar_file.cpp]: Would that also be the prefix on Windows?

Comment (mgudemann) [inline src/java_bytecode/jar_file.cpp]: apparently, yes, cf. https://en.wikipedia.org/wiki/WAR_(file_format)
what's currently missing though, is loading of jars listed in `WEB-INF/lib`

Comment (tautschnig) [inline src/java_bytecode/jar_file.cpp]: Ok, thanks!

Comment (tautschnig) [inline src/java_bytecode/jar_file.cpp]: (Maybe worth adding that link as a comment in the source code.)

File: src/java_bytecode/jar_file.cpp [hunk 4]
```diff
@@ -9,9 +9,9 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <cstring>
 #include <cassert>
 #include <unordered_set>
-
```
Comment (tautschnig) [inline src/java_bytecode/jar_file.cpp]: I'd rather have more blank lines that fewer - is there a reason (such as IDEs doing automatic work) why this line is gone?

Comment (mgudemann) [inline src/java_bytecode/jar_file.cpp]: re-added this line and one in front of documentation block for the function afterwards
no particular reason for having removed it

**Comment by mgudemann** [comment]
please note that this PR is not yet complete, as loading of jars in the war is not yet supported

**Comment by tautschnig** [comment]
Should it be tagged do-not-merge?

**Comment by mgudemann** [comment]
currently, yes, but I cannot do the tagging apparently

**Comment by mgudemann** [comment]
I brought this up to current `develop` but it still lacks support for .jar files in .war files. We probably could extract the packaged jars into temp files, load them and then delete them again? Or we can see how miniz supports extracting archives from memory. This will certainly require low-level API use.

**Comment by thomasspriggs** [comment]
I am going to close this PR as it appears that work on it has been inactive for an extended period of time. This is not any judgement on how worthwhile this work is. This is part of an effort to reduce the number of open PRs which are not being actively worked on. If you still think it would be worthwhile to get this merged then the branch should be re-based on the latest version of develop and the PR re-opened.

**Pull request closed without merging**